impl/src/main/java/io/kroxylicious/testing/kafka/common/KafkaClusterConfig.java

```diff
@@ -602,22 +604,26 @@ public ConfigHolder(Properties properties, Integer externalPort, Integer anonPor
             this.endpoint = endpoint;
             this.brokerNum = brokerNum;
             this.kafkaKraftClusterId = kafkaKraftClusterId;
-            this.roles = properties.getProperty("process.roles", BROKER_ROLE);
+        }
+
+        private String getRoles() {
+            return properties.getProperty("process.roles", BROKER_ROLE);
         }
 
         public boolean isBroker() {
-            return this.roles.contains(BROKER_ROLE);
+            return getRoles().contains(BROKER_ROLE);
         }
 
         public boolean isController() {
-            return this.roles.contains(CONTROLLER_ROLE);
+            return getRoles().contains(CONTROLLER_ROLE);
         }
     }
 
     /**
-     * The interface Kafka endpoints.
+     * Describes aspects of the KafkaCluster that are only material in a particular environment,
+     * for example after port allocation or creation of temporary directories.
      */
-    public interface KafkaEndpoints {
+    public interface MaterializedKafkaCluster {
```

Comment:
This class doesn't actually need to know about MaterializedKafkaCluster so think a pull-up to the top level is in order. It does need to know about Listener, so it should stand at the top too.  See my PR.